In [116]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [70]:
df = pd.read_csv('Dataset/train.csv')

In [71]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550068 entries, 0 to 550067
Data columns (total 12 columns):
User_ID                       550068 non-null int64
Product_ID                    550068 non-null object
Gender                        550068 non-null object
Age                           550068 non-null object
Occupation                    550068 non-null int64
City_Category                 550068 non-null object
Stay_In_Current_City_Years    550068 non-null object
Marital_Status                550068 non-null int64
Product_Category_1            550068 non-null int64
Product_Category_2            376430 non-null float64
Product_Category_3            166821 non-null float64
Purchase                      550068 non-null int64
dtypes: float64(2), int64(5), object(5)
memory usage: 50.4+ MB


In [72]:
df.head()

User_ID Product_ID Gender   Age  Occupation City_Category  \
0  1000001  P00069042      F  0-17          10             A   
1  1000001  P00248942      F  0-17          10             A   
2  1000001  P00087842      F  0-17          10             A   
3  1000001  P00085442      F  0-17          10             A   
4  1000002  P00285442      M   55+          16             C   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2               0                   3   
1                          2               0                   1   
2                          2               0                  12   
3                          2               0                  12   
4                         4+               0                   8   

   Product_Category_2  Product_Category_3  Purchase  
0                 NaN                 NaN      8370  
1                 6.0                14.0     15200  
2                 NaN                 NaN      1422  
3                14.0                 NaN      1057  
4                 NaN                 NaN      7969

In [77]:
#User_ID and Product_ID would not be used for training.. so delete them..
df_new = df.drop(['User_ID','Product_ID'], axis='columns')

In [78]:
print(df_new['Gender'].unique())

['F' 'M']


In [79]:
print(sorted(df_new['Occupation'].unique()))
df_new.groupby('Occupation')['Occupation'].count()

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]


Occupation
0     69638
1     47426
2     26588
3     17650
4     72308
5     12177
6     20355
7     59133
8      1546
9      6291
10    12930
11    11586
12    31179
13     7728
14    27309
15    12165
16    25371
17    40043
18     6622
19     8461
20    33562
Name: Occupation, dtype: int64

In [80]:
print(df_new['Age'].unique())
df_new['Age'].describe()

['0-17' '55+' '26-35' '46-50' '51-55' '36-45' '18-25']


count     550068
unique         7
top        26-35
freq      219587
Name: Age, dtype: object

In [81]:
print(df_new['City_Category'].unique())
print(df_new['City_Category'].describe())

['A' 'C' 'B']
count     550068
unique         3
top            B
freq      231173
Name: City_Category, dtype: object


In [82]:
print(df_new['Stay_In_Current_City_Years'].unique())
print(df_new['Stay_In_Current_City_Years'].describe())
print(df_new.groupby('Stay_In_Current_City_Years')['Stay_In_Current_City_Years'].count())

['2' '4+' '3' '1' '0']
count     550068
unique         5
top            1
freq      193821
Name: Stay_In_Current_City_Years, dtype: object
Stay_In_Current_City_Years
0      74398
1     193821
2     101838
3      95285
4+     84726
Name: Stay_In_Current_City_Years, dtype: int64


In [83]:
print(df_new['Marital_Status'].unique())
print(df_new['Marital_Status'].describe())
print(df_new.groupby('Marital_Status')['Marital_Status'].count())

[0 1]
count    550068.000000
mean          0.409653
std           0.491770
min           0.000000
25%           0.000000
50%           0.000000
75%           1.000000
max           1.000000
Name: Marital_Status, dtype: float64
Marital_Status
0    324731
1    225337
Name: Marital_Status, dtype: int64


In [84]:
print(df_new['Product_Category_1'].unique())
print(df_new['Product_Category_1'].describe())
print(df_new.groupby('Product_Category_1')['Product_Category_1'].count())

[ 3  1 12  8  5  4  2  6 14 11 13 15  7 16 18 10 17  9 20 19]
count    550068.000000
mean          5.404270
std           3.936211
min           1.000000
25%           1.000000
50%           5.000000
75%           8.000000
max          20.000000
Name: Product_Category_1, dtype: float64
Product_Category_1
1     140378
2      23864
3      20213
4      11753
5     150933
6      20466
7       3721
8     113925
9        410
10      5125
11     24287
12      3947
13      5549
14      1523
15      6290
16      9828
17       578
18      3125
19      1603
20      2550
Name: Product_Category_1, dtype: int64


In [85]:
print(sorted(df_new['Product_Category_2'].unique()))
print(df_new['Product_Category_2'].describe())
print(df_new.groupby('Product_Category_2')['Product_Category_2'].count())
df_new['Product_Category_2'].isnull().sum()

[nan, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0]
count    376430.000000
mean          9.842329
std           5.086590
min           2.000000
25%           5.000000
50%           9.000000
75%          15.000000
max          18.000000
Name: Product_Category_2, dtype: float64
Product_Category_2
2.0     49217
3.0      2884
4.0     25677
5.0     26235
6.0     16466
7.0       626
8.0     64088
9.0      5693
10.0     3043
11.0    14134
12.0     5528
13.0    10531
14.0    55108
15.0    37855
16.0    43255
17.0    13320
18.0     2770
Name: Product_Category_2, dtype: int64


173638

In [86]:
print(sorted(df_new['Product_Category_3'].unique()))
print(df_new['Product_Category_3'].describe())
print(df_new.groupby('Product_Category_3')['Product_Category_3'].count())
df_new['Product_Category_3'].isnull().sum()

[nan, 3.0, 4.0, 5.0, 6.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0]
count    166821.000000
mean         12.668243
std           4.125338
min           3.000000
25%           9.000000
50%          14.000000
75%          16.000000
max          18.000000
Name: Product_Category_3, dtype: float64
Product_Category_3
3.0       613
4.0      1875
5.0     16658
6.0      4890
8.0     12562
9.0     11579
10.0     1726
11.0     1805
12.0     9246
13.0     5459
14.0    18428
15.0    28013
16.0    32636
17.0    16702
18.0     4629
Name: Product_Category_3, dtype: int64


383247

In [87]:
print(df_new['Purchase'].describe())

count    550068.000000
mean       9263.968713
std        5023.065394
min          12.000000
25%        5823.000000
50%        8047.000000
75%       12054.000000
max       23961.000000
Name: Purchase, dtype: float64


## Initial Analysis of features and target
* Gender has two categorical values F and M. So we will have to convert Gender from Object data type to categorical
* Occupation has 21 unique categories from 0 to 21. So it should be converted to categorical.
* Age is divided into 7 unique intervals. So convert it into categorical.
* Three categories of citys are there. So convert to categorical.
* Stay_In_Current_City_Years has 5 unique values. So convert to categorical.
* Marital Status has two values 0 and 1. So convert to categorical.
* Product_Category_1 has 20 unique values. Convert to categorical.
* Product_Category_2 has 173638 values missing specified as NaN. These need to be rectified. It has 17 categorical.
* Product_Category_3 has 383247 values missing specified as NaN. These need to be rectified. It has 15 categorical values.
* Purchase is the target variable. None are missing.

Since there are many missing values in Product_Category_2 and Product_Category_3, it is not feasable to remove all those rows. 


In [91]:
# create a new column for category which is merged value of the three categories.
# First replace NaN with empty string and convert the column to string data type.
df_new['Product_Category_1'] = df_new['Product_Category_1'].astype(str)
df_new['Product_Category_2'] = df_new['Product_Category_2'].astype(str)
df_new['Product_Category_3'] = df_new['Product_Category_3'].astype(str)

In [96]:
#replace the string nan with empty string
df_new.loc[df_new['Product_Category_2'] == 'nan','Product_Category_2'] = '-'

In [97]:
df_new.loc[df_new['Product_Category_3'] == 'nan', 'Product_Category_3'] = '-'

In [98]:
print(df_new['Product_Category_2'].unique())
print(df_new['Product_Category_2'].describe())
print(df_new.groupby('Product_Category_2')['Product_Category_2'].count())

['-' '6.0' '14.0' '2.0' '8.0' '15.0' '16.0' '11.0' '5.0' '3.0' '4.0'
 '12.0' '9.0' '10.0' '17.0' '13.0' '7.0' '18.0']
count     550068
unique        18
top            -
freq      173638
Name: Product_Category_2, dtype: object
Product_Category_2
-       173638
10.0      3043
11.0     14134
12.0      5528
13.0     10531
14.0     55108
15.0     37855
16.0     43255
17.0     13320
18.0      2770
2.0      49217
3.0       2884
4.0      25677
5.0      26235
6.0      16466
7.0        626
8.0      64088
9.0       5693
Name: Product_Category_2, dtype: int64


In [99]:
print(df_new['Product_Category_3'].unique())
print(df_new['Product_Category_3'].describe())
print(df_new.groupby('Product_Category_3')['Product_Category_3'].count())

['-' '14.0' '17.0' '5.0' '4.0' '16.0' '15.0' '8.0' '9.0' '13.0' '6.0'
 '12.0' '3.0' '18.0' '11.0' '10.0']
count     550068
unique        16
top            -
freq      383247
Name: Product_Category_3, dtype: object
Product_Category_3
-       383247
10.0      1726
11.0      1805
12.0      9246
13.0      5459
14.0     18428
15.0     28013
16.0     32636
17.0     16702
18.0      4629
3.0        613
4.0       1875
5.0      16658
6.0       4890
8.0      12562
9.0      11579
Name: Product_Category_3, dtype: int64


Now creating a new feature from three other product types

In [107]:
df_new['final_category'] = df_new['Product_Category_1']+','+df_new['Product_Category_2']+','+df_new['Product_Category_3']

In [108]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550068 entries, 0 to 550067
Data columns (total 11 columns):
Gender                        550068 non-null object
Age                           550068 non-null object
Occupation                    550068 non-null int64
City_Category                 550068 non-null object
Stay_In_Current_City_Years    550068 non-null object
Marital_Status                550068 non-null int64
Product_Category_1            550068 non-null object
Product_Category_2            550068 non-null object
Product_Category_3            550068 non-null object
Purchase                      550068 non-null int64
final_category                550068 non-null object
dtypes: int64(3), object(8)
memory usage: 46.2+ MB


In [115]:
print(df_new['final_category'].unique())
print(df_new['final_category'].describe())
print(df_new.groupby('final_category')['final_category'].count())

['3,-,-' '1,6.0,14.0' '12,-,-' '12,14.0,-' '8,-,-' '1,2.0,-' '1,8.0,17.0'
 '1,15.0,-' '1,16.0,-' '5,11.0,-' '1,2.0,5.0' '5,8.0,14.0' '4,5.0,-'
 '2,3.0,4.0' '5,14.0,-' '1,14.0,16.0' '1,5.0,15.0' '1,8.0,14.0' '6,8.0,-'
 '2,4.0,8.0' '5,-,-' '3,4.0,-' '2,4.0,9.0' '5,12.0,-' '1,11.0,15.0'
 '14,-,-' '1,15.0,16.0' '5,9.0,-' '6,10.0,13.0' '1,6.0,8.0' '8,17.0,-'
 '8,13.0,-' '1,6.0,-' '5,8.0,-' '8,14.0,-' '1,5.0,6.0' '8,11.0,-'
 '1,8.0,-' '8,16.0,-' '1,2.0,6.0' '1,2.0,15.0' '1,6.0,15.0' '5,15.0,-'
 '1,15.0,17.0' '3,4.0,9.0' '4,8.0,9.0' '1,2.0,13.0' '2,3.0,15.0'
 '2,9.0,14.0' '3,4.0,5.0' '5,6.0,-' '5,6.0,13.0' '1,-,-' '3,5.0,-'
 '11,-,-' '3,4.0,12.0' '1,2.0,4.0' '3,12.0,-' '1,11.0,16.0' '2,4.0,15.0'
 '1,2.0,3.0' '6,16.0,-' '11,16.0,-' '1,6.0,16.0' '2,8.0,14.0' '13,16.0,-'
 '1,5.0,12.0' '1,5.0,9.0' '2,4.0,-' '5,8.0,18.0' '1,8.0,9.0' '2,-,-'
 '2,17.0,-' '8,14.0,17.0' '1,5.0,-' '15,-,-' '8,15.0,-' '8,10.0,-'
 '5,9.0,14.0' '7,-,-' '8,13.0,15.0' '16,-,-' '1,8.0,16.0' '4,5.0,9.0'
 '1,2.0,14.0' '5,17.0,

New product category 'final_category' has a total of 237 category types which could be converted to categorical values.

In [118]:
df_final = df_new.drop(['Product_Category_1','Product_Category_2','Product_Category_3'], axis='columns')

In [119]:
df_final.head()

Gender   Age  Occupation City_Category Stay_In_Current_City_Years  \
0      F  0-17          10             A                          2   
1      F  0-17          10             A                          2   
2      F  0-17          10             A                          2   
3      F  0-17          10             A                          2   
4      M   55+          16             C                         4+   

   Marital_Status  Purchase final_category  
0               0      8370          3,-,-  
1               0     15200     1,6.0,14.0  
2               0      1422         12,-,-  
3               0      1057      12,14.0,-  
4               0      7969          8,-,-

## Now doing some EDA